sql/2024/04/image-formats.sql

```diff
@@ -0,0 +1,87 @@
+CREATE TEMPORARY FUNCTION
+  IS_CMS(technologies ARRAY<STRUCT<technology STRING,
+                                   categories ARRAY<STRING>,
+                                   info ARRAY<STRING>>>,
+         cms STRING,
+         version STRING)
+  RETURNS BOOL AS ( EXISTS(
+  SELECT
+    *
+  FROM
+    UNNEST(technologies) AS technology,
+    UNNEST(technology.info) AS info
+  WHERE
+      technology.technology = cms
+    AND ( version = ""
+    OR ENDS_WITH(version, ".x")
+            AND (STARTS_WITH(info, RTRIM(version, "x"))
+        OR info = RTRIM(version, ".x"))
+    OR info = version ) ) );
+
+WITH
+  pagesWithImages AS (
+    SELECT
+      date,
+      client,
+      page,
+      LOWER(JSON_EXTRACT_SCALAR(custom_metrics, '$.performance.lcp_elem_stats.nodeName')) = 'img' AS has_lcp_image,
+      LOWER(JSON_EXTRACT_SCALAR(custom_metrics, '$.performance.lcp_elem_stats.url')) = JSON_EXTRACT_SCALAR(images, '$.url') AS is_lcp_image,
```

Comment:
I believe the query is only interested in LCP images, right? In that case, I think we should put these two things into the `WHERE` clause instead so that only LCP images are included in the result of the `pagesWithImages` query. It probably also reduces the data processed drastically, which should help reduce the cost of the `JOIN`.

Reply:
As discussed last time and mentioned in my summary above, we _could_ expand this to all images on pages. ıut yeah that's a lot of data, and for the sake of this analysis I think LCP is good enough.